### ISSUE TITLE
Connection's filter list isn't always updated

### ISSUE THREAD

**Issue by soininen** [Spine-project/Spine-Toolbox#1796]
We can't rely on the database manager to keep connection filters up-to-date with scenarios/tools because executing items like Importer or Merger or having "Purge before writing" set won't trigger any signals in the manager.

Steps to reproduce:
1. Have a Data store without scenarios
2. Check that outgoing links from the DS don't have filters
3. Merge scenarios from another database to the DS
4. Check the outgoing links from DS again - they don't list the merged scenarios as they should

**Comment by soininen** [comment]
Actually, we can rely on the database manager as long connections also receive notifications about commits and rollbacks. The above commits also fix some issues connected to database workers outliving the database editor and purging the database from Data store's properties tab.

**Comment by soininen** [comment]
There seems to be more work to be done on this. Thanks for the code review, @manuelma!

**Comment by manuelma** [comment]
No worries - excellent work so far fixing my mistakes.

**Comment by soininen** [comment]
I feel we're heading in the right direction with using the Database manager in other parts of Toolbox than just Database editor. I'm experimenting with using Database manager in the View item in my plotting rewrite branch.

**Comment by soininen** [comment]
Fixed issues found in code review.

**Comment by manuelma** [comment]
@soininen there are still a couple of minor issues we can fix. I hinted at them in comments within https://github.com/Spine-project/Spine-Toolbox/commit/759103936ffc5c96b2651622a84146049d21d0c3 but I will just list them here for convenience. You can reproduce them using this 'large' DB: [OEO.zip](https://github.com/Spine-project/Spine-Toolbox/files/9823470/OEO.zip)
1. `SpineDBWorker.can_fetch_more()` sometimes freezes the UI. In my machine the editor is painted black for a few seconds when loading the above DB. I believe one solution is to just return True if the query is not initialized and let `fetch_more()` deal with the consequences.
2. There might be infinite fetch loops. With the above DB, I believe it happens if one expands the regions__comm__tech__vintage__comm___emissions class in the Relationship tree for instance. I believe the solution is to simple restore the `QTimer.singleShot()` to mark the fetch parent as not busy in `SpineDBWorker._fetch_event`.

**Comment by soininen** [comment]
Alright, I'll have a look. Thanks for the database, it's always nice to have new test cases!

**Comment by soininen** [comment]
I fixed the infinite fetching issue (number 2 in the post above) for now but `can_fetch_more()` needs more work. The current approach using `QWaitCondition` results in a deadlock in unit tests on Linux proving that the implementation is not as fool proof as I thought.

On the other hand, if we just ignored the case where the query wasn't initialized, unit tests might get stuck as well which made me implement the `QWaitCondition` approach in the first place.

Blocking on futures makes the UI unresponsive which is no-go. I'll see if I can improve on this by calling `QApplication.processEvents()` while the worker thread is initializing the query.

**Comment by manuelma** [comment]
Thanks for looking at this @soininen 
The QApplication.processEvents() might work, good idea. In that case we can apply it so we never return from can_fetch_more before we know the result. And in that case, parent.fetch_status_change() can also be removed.

**Comment by soininen** [comment]
I ended up returning to the previous status quo with no waiting in `can_fetch_more()` at all. Using futures + `QApplication.processEvents()` resulted in crashes similar to [here](https://stackoverflow.com/questions/67772456/process-finished-with-exit-code-107341571-0xc00000fd). Perhaps we ended up calling `processEvents()` recursively too many times?

In any case, DB editor seems to work and is somewhat responsive with @manuelma's test database. Unit tests get stuck on my Windows system sometimes but that is just the case with the status quo. Also, I observed some Tracebacks which are probably unrelated, but I'll still have a look.
